### PR TITLE
fix: fake all supported timers by default

### PR DESCRIPTION
**Pull request by SimenB**
#### Purpose (TL;DR) - mandatory
It's very weird to me that this method is excluded from default mocking. I've read #126 and to me the solution seems to be for users to specify a `toFake` themselves, not force other people to specify "yes, when I say fake timers I do mean _all_ timers".

Note that this should be considered a breaking change.

Whenever we start supporting some _new_ timer, it should by default not be mocked, but I think when we later make a new major release it should be mocked by default again.

### CODE REVIEW

File: src/fake-timers-src.js [hunk 1]
```diff
@@ -1262,10 +1262,7 @@ function withGlobal(_global) {
         clock.methods = config.toFake || [];
 
         if (clock.methods.length === 0) {
-            // do not fake nextTick by default - GitHub#126
-            clock.methods = keys(timers).filter(function(key) {
-                return key !== "nextTick" && key !== "queueMicrotask";
```
Comment (SimenB) [inline src/fake-timers-src.js:1267]: seems docs were lying 🤷

File: src/fake-timers-src.js [hunk 2]
```diff
@@ -1259,14 +1259,7 @@ function withGlobal(_global) {
...
 
-        clock.methods = config.toFake || [];
-
-        if (clock.methods.length === 0) {
-            // do not fake nextTick by default - GitHub#126
-            clock.methods = keys(timers).filter(function(key) {
-                return key !== "nextTick" && key !== "queueMicrotask";
-            });
-        }
+        clock.methods = config.toFake || keys(timers);
```
Comment (SimenB) [inline src/fake-timers-src.js:1262]: another technically breaking change I guess - if people passed `toFake: []` before, we faked everything. Seems like a bug fix to me to respect it, but maybe not 😀 If _not_, should just revert the last commit

**Comment by benjamingr** [comment]
So here's the thing: promise libraries and other libraries use nextTick and microticks for scheduling so this is a pretty big breaking change.

In principle I agree that we should mock this by default.

File: test/fake-timers-test.js [hunk 3]
```diff
@@ -4504,19 +4504,6 @@ describe("FakeTimers", function() {
             assert(called);
             clock.uninstall();
         });
-
-        it("does not install by default - GitHub#126", function(done) {
```
Comment (benjamingr) [inline test/fake-timers-test.js:4508]: Would it be possible to change this test to assert nextTick is mocked if the behavior changes?

File: src/fake-timers-src.js (hunk 2 above)
Comment (benjamingr) [reply src/fake-timers-src.js:1262]: Yeah I think this and the other one should be a major version if we land them.

**Comment by SimenB** [comment]
> So here's the thing: promise libraries and other libraries use nextTick and microticks for scheduling so this is a pretty big breaking change.

Good point, should probably call it out in the readme.

> In principle I agree that we should mock this by default.

👍

File: test/fake-timers-test.js (hunk 3 above)
Comment (SimenB) [reply test/fake-timers-test.js:4508]: changed the test above, I think that's fine?

**Comment by SimenB** [comment]
@benjamingr added a section to the readme, feedback appreciated 😀

File: src/fake-timers-src.js (hunk 2 above)
Comment (SimenB) [reply src/fake-timers-src.js:1262]: Can open a separate PR for it?

**Comment by benjamingr** [comment]
@SimenB I am +1 on this change in principle but I know it will break pretty much every fake-timers install "in the wild" that uses the current API.

**Comment by SimenB** [comment]
Should work if people use the builtin `Promise` and `async-await`, no? If it breaks builtin `Promise` global I'd consider that a bug in node.

My main issue is that to enable "fake everything" I have to enumerate all timers from the outside, which is not obvious I think. Maybe we could rather add a `fakeNextTick` explicit option (defaulting to `false` I guess) instead of piggy-backing `toFake`?

**Comment by benjamingr** [comment]
> Should work if people use the builtin Promise and async-await, no? If it breaks builtin Promise global I'd consider that a bug in node.

It doesn't break the built in promises but mostly because we don't mock its timings at the moment (we probably should with async_hooks). In practice this means that it's a lot harder than it has to to test promises with fake-timers (I recall a lot of discussions about this + async versions of timers somewhere?)

I don't mind seeing what breaks with mocking nextTick on some large codebases I maintain - though I assume I know the answer is "every place that relies on nextTick executing and doesn't .tick or runMicrotasks before that :]

**Comment by SimenB** [comment]
In a work project where I migrated to use Jest's fake timers now that Jest ships with Lolex it broke `node-fetch` (and/or integration with `nock`, not entirely sure)

**Comment by mroderick** [comment]
Stale bot?!

**Comment by fatso83** [comment]
This makes my head hurt and you guys know best anyhow. Not sure why it has stalled, though?

**Review by fatso83 — APPROVED** [review]
To me, this looks good. @SimenB anything you want addressed (apart from the conflicts)?

**Comment by fatso83** [comment]
B,B,B,B,bump? It's not super problematic to release a major version, as long as you also document how to deal with the fallout. Like using node-fetch as an example case, show what needs to be done to fix the test cases. @SimenB

File: src/fake-timers-src.js (hunk 2 above)
Comment (fatso83) [reply src/fake-timers-src.js:1262]: > Can open a separate PR for it?

Yes

File: test/fake-timers-test.js (hunk 3 above)
Comment (fatso83) [reply test/fake-timers-test.js:4508]: > changed the test above, I think that's fine?

Yes, this is fine. It's implicitly true, as the code won't work unless `nextTick` is synchronous.

**Review by fatso83 — APPROVED** [review]

**Comment by mroderick** [comment]
This needs a rebase to resolve the conflicts.

@SimenB do you have time to take this across the finish line? If not, would you be ok with someone else doing it?

**Comment by fatso83** [comment]
**I want to finish this (and do #490) and release a new breaking major. Soonish** 

> I don't mind seeing what breaks with mocking nextTick on some large codebases I maintain

There's been 4 years. I guess the number of places that rely on non-native Promises are a lot fewer and further between. Maybe it's not that bad? I think we would just need to add something to the changelog.

**Comment by fatso83** [comment]
Since I closed #490 and I haven't heard anything in months, I think I'll just pick this up and release a new breaking major.

**Comment by fatso83** [comment]
Weird how much you can get done in a weekend without kids 😄

**Comment by fatso83** [comment]
I think this needs to be reverted. It seems to trip people up over and over again, ref https://github.com/sinonjs/sinon/issues/2620#issuecomment-2461063710, and so it makes for poor developer experience. `nextTick` and `queueMicroTask` is far less common to target than `setTimeout` and similar, but it's the latter crowd that now needs to supply a list of timers, instead of the more exotic crowd.

I'd rather just change the docs to say "fakes _most_ timers by default, but you can also choose to stub all or a subset, if you so please" and then expose some constants they could use (`DEFAULT_TIMERS`,`ES_STANDARD_TIMERS`, `ALL_TIMERS`), a la:
```javascript
import 
fakeTimers.install({toFake: fakeTimers.ALL_TIMERS})
```

**Pull request merged**